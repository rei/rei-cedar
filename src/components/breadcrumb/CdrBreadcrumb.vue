<script>
import {
  defineComponent, useCssModule, computed, ref, watch, nextTick,
} from 'vue';
import uid from '../../utils/uid';

/** Navigation used to reveal a page's location within the site hierarchy */
export default defineComponent({
  name: 'CdrBreadcrumb',
  props: {
    /**
     * Sets the array of a breadcrumb object containing a 'url' and 'name' property.
     * @demoIgnore true
     */
    items: {
      type: Array,
      default: () => [],
      validator: (value) => {
        if (value.length && value.length > 0) {
          for (let i = 0; i < value.length; i += 1) {
            if (!(typeof value[i].item === 'object')) {
              console.error('Breadcrumb items array missing item key at index ', i); // eslint-disable-line no-console
              return false;
            }
            if (!Object.hasOwnProperty.call(value[i].item, 'name')) {
              console.error('Breadcrumb items array is missing item.name value at index ', i); // eslint-disable-line no-console
              return false;
            }
          }
        }
        return true;
      },
    },
    /**
     * Controls the ability to truncate the entire breadcrumb path. If this value is false, truncation will no longer occur.
     */
    truncationEnabled: {
      type: Boolean,
      default: true,
    },
    /**
     * Define a custom ID for the `<nav>` element. Randomly generated if no ID provided.
     * @demoIgnore true
     */
    id: {
      type: String,
    },
  },
  emits: {
    /**
     * Emits when a breadcrumb item is clicked. `e.preventDefault()` may be used to override the default link navigation.
     */
    navigate: null,
  },

  setup(props) {
    const uniqueId = props.id ? props.id : uid();
    const truncate = ref(props.truncationEnabled && props.items.length > 2);
    const itemListEl = ref(null);
    const ellipsisLabel = computed(() => {
      const s = (props.items.length - 2) > 1 ? 's' : '';
      return `show ${props.items.length - 2} more navigation level${s}`;
    });

    const handleEllipsisClick = () => {
      truncate.value = false;
      nextTick(() => {
        itemListEl.value.querySelector('li a').focus();
      });
    };

    watch(() => props.items, () => {
      truncate.value = props.truncationEnabled && props.items.length > 2;
    });

    return {
      style: useCssModule(),
      uniqueId,
      truncate,
      itemListEl,
      ellipsisLabel,
      handleEllipsisClick,
    };
  },
});
</script>

<template>
  <nav
    :class="style['cdr-breadcrumb']"
    :id="uniqueId"
    aria-label="breadcrumbs"
  >
    <ol
      :id="`${uniqueId}List`"
      :class="style['cdr-breadcrumb__list']"
      ref="itemListEl"
    >
      <li
        :class="style['cdr-breadcrumb__item']"
        v-if="truncate"
      >
        <button
          @click="handleEllipsisClick"
          aria-expanded="false"
          :class="style['cdr-breadcrumb__ellipses']"
          :aria-controls="`${uniqueId}List`"
          :aria-label="ellipsisLabel"
        >
          <span
            :class="style['cdr-breadcrumb__ellipses-icon']"
            aria-hidden="true"
          >
            . . .
          </span>
        </button>
        <span
          :class="style['cdr-breadcrumb__delimiter']"
          aria-hidden="true"
        >
          /
        </span>
      </li>

      <li
        v-for="(breadcrumb, index) in items"
        :class="style['cdr-breadcrumb__item']"
        :key="breadcrumb.item.id || breadcrumb.item.name.replace(/ /g, '-').toLowerCase()"
        v-show="!truncate || (index >= items.length - 2)"
      >
        <slot
          name="link"
          :class="style['cdr-breadcrumb__link']"
          :href="breadcrumb.item.url"
          :content="breadcrumb.item.name"
        >
          <a
            :class="style['cdr-breadcrumb__link']"
            :href="breadcrumb.item.url"
            @click="$emit('navigate', breadcrumb, e)"
          >
            {{ breadcrumb.item.name }}
          </a>
        </slot>

        <span
          v-if="index < items.length - 1"
          :class="style['cdr-breadcrumb__delimiter']"
          aria-hidden="true"
        >
          /
        </span>
      </li>
    </ol>
  </nav>
</template>

<style lang="scss" module src="./styles/CdrBreadcrumb.module.scss">
</style>
