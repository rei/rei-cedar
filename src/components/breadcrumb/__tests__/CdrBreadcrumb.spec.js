import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrBreadcrumb from '../CdrBreadcrumb.vue';

const itemsA = [
  {
    item: {
      url: 'http://google.com',
      name: 'Longer Breadcrumb List Step 1',
    },
  },
  {
    item: {
      url: 'http://rei.com',
      name: 'Longer Breadcrumb Step 2',
    },
  },
  {
    item: {
      url: 'http://yahoo.com',
      name: 'Longer Breadcrumb Step 3',
    },
  },
];

const itemsB = [
  {
    item: {
      url: 'http://google.com',
      name: 'Shorter Breadcrumb List Step 1',
    },
  },
  {
    item: {
      url: 'http://rei.com',
      name: 'Shorter Breadcrumb List Step 2',
    },
  },
];

const itemsB2 = [
  {
    item: {
      url: 'http://google.com',
      name: 'Breadcrumb Extension Step 3',
    },
  },
  {
    item: {
      url: 'http://rei.com',
      name: 'Breadcrumb Extension Step 4',
    },
  },
];

describe('CdrBreadcrumb', () => {
  describe('with 3 or more breadcrumb items', () => {
    let wrapper;
    let ellipse;
    beforeEach(() => {
      wrapper = mount(CdrBreadcrumb, {
        propsData: {
          id: 'bc-test',
          items: itemsA
        },
        attachTo: document.body
      })
      ellipse = wrapper.find('.cdr-breadcrumb__ellipses');
    })
    it('renders correctly', async () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should truncate with 3 or more items', async () => {
      expect(wrapper.find('.cdr-breadcrumb__ellipses').exists()).toBe(true);
    });

    describe('a11y requirements', () => {
      it('the breadcrumb is a "NAV" element', () => {
        expect(wrapper.element.tagName).toBe('NAV');
      });

      it('the breadcrumb nav has the expected "aria-label" value', () => {
        expect(wrapper.attributes()['aria-label']).toBe('breadcrumbs');
      });

      it('the ellipse button has the expected "aria-label" value', () => {
        expect(ellipse.attributes()['aria-label']).toBe('show 1 more navigation level');
      });

      it('the ellipse button has the expected "aria-controls" value', () => {
        expect(ellipse.attributes()['aria-controls']).toBe('bc-testList');
      });

      it('the ellipse button has the expected "aria-expanded" value', () => {
        expect(ellipse.attributes()['aria-expanded']).toBe('false');
      });
    })

    describe('when truncated breadcrumb items list has been expanded', () => {
      beforeEach(() => {
        wrapper.find('.cdr-breadcrumb__ellipses').trigger('click');
      });

      it('is no longer truncated', () => {
        expect(wrapper.find('.cdr-breadcrumb__ellipses').exists()).toBeFalsy();
      });

      it('applies focus to first breadcrumb on ellipsis click', ()=>{
        expect(document.activeElement.textContent).toBe('Longer Breadcrumb List Step 1')
      })
    })
  });

  describe('when there is fewer than 3 items', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrBreadcrumb, {
        propsData: {
          id: 'bc-test',
          items: itemsB,
        }
      });
    });
    it('is not truncated', () => {
      expect(wrapper.find('.cdr-breadcrumb__ellipses').exists()).toBe(false);
    });

    it('breadcrumb should evaluate truncation when items are updated', async () => {
      expect(wrapper.find('.cdr-breadcrumb__ellipses').exists()).toBe(false);
      await wrapper.setProps({ items: itemsB.concat(itemsB2) });
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.cdr-breadcrumb__ellipses').exists()).toBe(true);
    });
  })
});