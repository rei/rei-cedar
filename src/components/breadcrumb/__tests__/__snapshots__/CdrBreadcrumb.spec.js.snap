// Vitest Snapshot v1

exports[`CdrBreadcrumb > with 3 or more breadcrumb items > renders correctly 1`] = `
<nav
  aria-label="breadcrumbs"
  class="cdr-breadcrumb"
  id="bc-test"
>
  <ol
    class="cdr-breadcrumb__list"
    id="bc-testList"
  >
    <li
      class="cdr-breadcrumb__item"
    >
      <button
        aria-controls="bc-testList"
        aria-expanded="false"
        aria-label="show 1 more navigation level"
        class="cdr-breadcrumb__ellipses"
      >
        <span
          aria-hidden="true"
          class="cdr-breadcrumb__ellipses-icon"
        >
           . . . 
        </span>
      </button>
      <span
        aria-hidden="true"
        class="cdr-breadcrumb__delimiter"
      >
         / 
      </span>
    </li>
    
    <li
      class="cdr-breadcrumb__item"
      style="display: none;"
    >
      
      <a
        class="cdr-breadcrumb__link"
        href="http://google.com"
      >
        Longer Breadcrumb List Step 1
      </a>
      
      <span
        aria-hidden="true"
        class="cdr-breadcrumb__delimiter"
      >
         / 
      </span>
    </li>
    <li
      class="cdr-breadcrumb__item"
    >
      
      <a
        class="cdr-breadcrumb__link"
        href="http://rei.com"
      >
        Longer Breadcrumb Step 2
      </a>
      
      <span
        aria-hidden="true"
        class="cdr-breadcrumb__delimiter"
      >
         / 
      </span>
    </li>
    <li
      class="cdr-breadcrumb__item"
    >
      
      <a
        class="cdr-breadcrumb__link"
        href="http://yahoo.com"
      >
        Longer Breadcrumb Step 3
      </a>
      
      <!--v-if-->
    </li>
    
  </ol>
</nav>
`;
